ambry-router/src/main/java/com/github/ambry/router/PutOperation.java

```diff
@@ -1428,8 +1469,15 @@ int fillFrom(ByteBuf channelReadBuf) {
         }
       }
       if (buf.readableBytes() == routerConfig.routerMaxPutChunkSizeBytes) {
-        onFillComplete(true);
-        updateChunkFillerWaitTimeMetrics();
+        if (firstPutChunkPrepared) {
+          onFillComplete(true);
+        } else {
+          // If first put chunk is full, but not yet prepared then mark it pending resolution instead of completing it.
+          // This chunk will be prepared as soon as either more bytes are read from the channel, or the chunk filling
+          // is complete. At that point we will have enough information to mark this blob as simple or composite blob.
+          chunkToFill = null; // this will ensure that a new chunk is picked up for filling more data.
```

Comment:
we probably don't ned to assign null to chunkToFill. When there are extra bytes from http channel, before we read any of bytes, we would already prepare first chunk to send it out. After that, the state of first chunk would be either encrypting or ready.